### DISCUSSION TITLE
Abstract the SMS notification channel

### DISCUSSION THREAD

**Discussion by Muffinman** [laravel/framework#50889] [Ideas]
Why are all the SMS notification methods all named `Vonage*`? [[docs]](https://laravel.com/docs/11.x/notifications#sms-notifications)

It seems a bit strange given that there are many SMS providers (e.g. Twilio/SendGrid) who all follow the same standards, and other similar services (e.g. search / mail) are all abstracted not to be service specific.
 
Having to use specific methods like `toVonage()` seems like it may be a bit of an anti-pattern.

Bad?
```php
use Illuminate\Notifications\Messages\VonageMessage;

/**
 * Get the notification's delivery channels.
 *
 * @return array<int, string>
 */
public function via(object $notifiable): array
{
   return ['vonage'];
}

/**
 * Get the Vonage / SMS representation of the notification.
 */
public function toVonage(object $notifiable): VonageMessage
{
    return (new VonageMessage)
                ->content('Your SMS message content');
}
```

Better?

```
SMS_DRIVER=vonage/twilio/sendgrid
```

```php
use Illuminate\Notifications\Messages\SMSMessage;

/**
 * Get the notification's delivery channels.
 *
 * @return array<int, string>
 */
public function via(object $notifiable): array
{
   return ['sms'];
}

/**
 * Get the SMS representation of the notification.
 */
public function toSMS(object $notifiable): SMSMessage
{
    return (new SMSMessage)
                ->content('Your SMS message content');
}
```

**Comment by JohnRoux** [comment]
I've added in several custom laravel sms notification channels in my time.

The real question I have, is why don't we add some contracts to the channels!
Currently everything is just assumed and you hope it works.
Having a `NotificationChannelContract` with a `send` method would make a lot of sense!


But yeah, on the sms stuff you normally also want delivery report tracking as well.
I normally add a contract for `ShouldTrackSmsDelivery` to my notifications which can send via sms and have an event listener do some handling for the tracking side

**Comment by JohnRoux** [comment]
> It seems a bit strange given that there are many SMS providers (e.g. Twilio/SendGrid) who all follow the same standards, and other similar services (e.g. search / mail) are all abstracted not to be service specific.


Yeeeeeah, I've seen a LOT of SMS providers who also _don't_ follow those same standards.
But having some sort of SMS management would still be good, as long as custom providers can still slot in nicely

**Comment by Muffinman** [comment]
Coming back to this over a year later, and I now realise that this Vonage stuff is all already correctly separated into it's own package.

The confusion on my part was merely that [the package](https://github.com/laravel/vonage-notification-channel/) uses a namespace of `Illuminate\Notifications\Messages`, such that looks like part of the core framework.
